datafusion/src/datasource/object_store/mod.rs

```diff
@@ -20,54 +20,68 @@
 pub mod local;
 
 use std::collections::HashMap;
-use std::fmt::Debug;
+use std::fmt::{self, Debug};
+use std::io::Read;
 use std::pin::Pin;
 use std::sync::{Arc, RwLock};
 
 use async_trait::async_trait;
-use futures::{AsyncRead, Stream};
+use futures::{Stream, StreamExt};
 
 use local::LocalFileSystem;
 
 use crate::error::{DataFusionError, Result};
-use chrono::Utc;
 
-/// Object Reader for one file in a object store
+/// Object Reader for one file in an object store
+/// Note that the dynamic dispatch on the reader might
+/// have some performance impacts.
 #[async_trait]
 pub trait ObjectReader {
-    /// Get reader for a part [start, start + length] in the file asynchronously
-    async fn chunk_reader(&self, start: u64, length: usize)
-        -> Result<Arc<dyn AsyncRead>>;
+    /// Get reader for a part [start, start + length] in the file
+    fn chunk_reader(
+        &self,
+        start: u64,
+        length: usize,
+    ) -> Result<Box<dyn Read + Send + Sync>>;
+
+    /// Get reader for the entire file
+    fn reader(&self) -> Result<Box<dyn Read + Send + Sync>> {
+        self.chunk_reader(0, self.length() as usize)
+    }
 
-    /// Get length for the file
+    /// Get the size of the file
     fn length(&self) -> u64;
 }
 
 /// Represents a file or a prefix that may require further resolution
 #[derive(Debug)]
 pub enum ListEntry {
-    /// File metadata
-    FileMeta(FileMeta),
+    /// Complete file path with size
+    SizedFile(SizedFile),
     /// Prefix to be further resolved during partition discovery
     Prefix(String),
 }
 
-/// File meta we got from object store
-#[derive(Debug)]
-pub struct FileMeta {
+/// Complete file path with size we got from object store
+#[derive(Debug, Clone)]
+pub struct SizedFile {
     /// Path of the file
     pub path: String,
-    /// Last time the file was modified in UTC
-    pub last_modified: Option<chrono::DateTime<Utc>>,
```

Comment:
Sorry for that, just an oversight. I'll look how it is used exactly and add it back with some comments 🙂

Reply:
I tried to understand what `last_modified` is used for exactly in the delta reader, but I couldn't quite get it... According to my understanding, the modification dates that should be used are sourced from the catalog (for example [this log file](https://github.com/delta-io/delta-rs/blob/main/rust/tests/data/delta-0.8.0/_delta_log/00000000000000000000.json)), not the object store.

Reply:
The main use-case in delta is to get the timestamp for a specific verison commit. The commit entry you linked contains timestamp metadata for each individual action, but not the full commit/version. The object/commit timestamp is currently only used in one place: time travel. Here is the relevant code: https://github.com/delta-io/delta-rs/blob/e52f23fd3ac68b86aededff081dffcd86902ecf8/rust/src/delta.rs#L809. The idea being we can leverage object list and head API calls to perform version filtering by timestamp efficiently without having to fetch and parse the object content.

Reply:
Thanks @houqp for these pointers, I understand better now. I am really surprised by that behavior 😃. This basically means you cannot copy a delta table around (https://github.com/delta-io/delta/issues/192). I will add back this field right away with a pointer to this issue. 

Out of curiosity. I can't find any reference to the usage of the file system / object store timestamp in the Delta [documentation](https://docs.delta.io/latest/index.html) or the [protocol](https://github.com/delta-io/delta/blob/master/PROTOCOL.md). Have I missed something? If I had designed this, I would have definitively encoded the timestamp in the log entry key (something like `00000000000000000000-1615043767813.json`) 😄

Reply:
That's a good point, yeah, this design is a bit hacky :D This behavior is not documented in the official doc nor the protocol spec, I think I probably got it from reading the scala reference implementation.

I think the reason they went with `version.json` instead of `version-ts.json` was so that they can easily fetch version commit object directly with get api call without having to do a list api call first. But I agree with you that it's not ideal to depend on object modification time in this particular design, tracking the version timestamp in a separate managed metadata folder would have been better. For example, a version timestamp checkpoint file or some sort.